dipy/direction/probabilistic_direction_getter.pyx

```diff
@@ -93,14 +93,23 @@ cdef class ProbabilisticDirectionGetter(PmfGenDirectionGetter):
         cdef:
             cnp.npy_intp i, idx, _len
             double[:] newdir, pmf
+            double vertex[3]
             double last_cdf, random_sample
             cnp.uint8_t[:] bool_array
 
         pmf = self._get_pmf(point)
         _len = pmf.shape[0]
-
-        bool_array = self._adj_matrix[
-            (direction[0], direction[1], direction[2])]
+        try:
+            bool_array = self._adj_matrix[
+                (direction[0], direction[1], direction[2])]
+        except KeyError:
+            # if direction is not a sphere vertex, we find the closest vertex
+            for i in range(3):
+                vertex[i] = direction[i]
+            idx = self.sphere.find_closest(vertex)
+            bool_array = self._adj_matrix[(self.vertices[idx, :][0],
+                                           self.vertices[idx, :][1],
+                                           self.vertices[idx, :][2])]
```

Comment:
Not sure about this.  `try except` python clause inside a `cdef` functions. 

You might loose all the cedf benefit.  I need to look deeper at this.

Reply:
I just realized that `self._adj_matrix` is a dict, so a python object.

This needs to change. otherwise the `cdef` is quite useless.

Reply:
Replacing the `dict` with a cythonized search for neighbors vertices turn out as fast for probabilistic tracking and faster for deterministic maximum tracking. The code is shorter and more clear this way. Further optimization will be done in another PR.

Reply:
That's great! Thank you @gabknight. waiting for the CI's to finish and I will go ahead to merge this